File: scala/src/main/scala/com/authdog/types/Meta.scala
package com.authdog.types

import io.circe.{Decoder, Encoder, HCursor, Json}
import io.circe.syntax._

/**
 * Metadata in the response
 */
case class Meta(
  code: Int,
  message: String
)

object Meta {
  implicit val decoder: Decoder[Meta] = (c: HCursor) => for {
    code <- c.downField("code").as[Int]
    message <- c.downField("message").as[String]
  } yield Meta(code, message)
  
  implicit val encoder: Encoder[Meta] = (meta: Meta) => Json.obj(
    "code" -> meta.code.asJson,
    "message" -> meta.message.asJson
  )
}

File: java/src/main/java/com/authdog/types/Meta.java
package com.authdog.types;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Metadata in the response.
 */
public final class Meta {
    /**
     * Response code.
     */
    @JsonProperty("code")
    private int code;

    /**
     * Response message.
     */
    @JsonProperty("message")
    private String message;

    /**
     * Default constructor.
     */
    public Meta() {
    }

    /**
     * Constructor with parameters.
     * @param codeParam Response code
     * @param messageParam Response message
     */
    public Meta(final int codeParam, final String messageParam) {
        this.code = codeParam;
        this.message = messageParam;
    }

    /**
     * Get the response code.
     * @return Response code
     */
    public int getCode() {
        return code;
    }

    /**
     * Set the response code.
     * @param codeParam Response code
     */
    public void setCode(final int codeParam) {
        this.code = codeParam;
    }

    /**
     * Get the response message.
     * @return Response message
     */
    public String getMessage() {
        return message;
    }

    /**
     * Set the response message.
     * @param messageParam Response message
     */
    public void setMessage(final String messageParam) {
        this.message = messageParam;
    }
}
